### PR TITLE
arch: Add Atmel SAM4S SoC support

### PR DESCRIPTION
**Pull request by fallrisk**
Added support for Atmel SAM4S (Cortex-M4) MCU:
- Kconfig files
- Devicetree files

Signed-off-by: Justin Watson <jwatson5@gmail.com>

### CODE REVIEW

File: arch/arm/soc/atmel_sam/sam4s/linker.ld [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/* linker.ld - Linker command/script file */
+
+/*
```
Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/linker.ld]: /*
 * Copyright (c) 2014 Wind River Systems, Inc.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

File: arch/arm/soc/atmel_sam/sam4s/soc.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/soc.c]: Replace copyright header with SPDX tag

/*
 *
 * SPDX-License-Identifier: Apache-2.0
 */

Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/soc.c]: I meant license, not copyright

File: arch/arm/soc/atmel_sam/sam4s/soc.c [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	key = irq_lock();
+
+	/* Setup the flash controller.
+	 * The bootloader is running @ 48 MHz with
+	 * FWS == 2.
+	 * When running at 84 MHz, FWS == 4 seems
+	 * to be more stable, and allows the board
+	 * to boot.
+	 */
+	__EEFC0->fmr = 0x00000400;
```
Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/soc.c]: can we use some #defines instead of magic number

File: dts/arm/atmel/sam4s.dtsi [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		reg = <0x20100000 0x20000>;
+	};
+
+	flash0: flash {
+		compatible = "flash";
+		reg = <0x00400000 0x100000>;
+	};
+
+	soc {
+		uart0: uart@400E0800 {
```
Comment (galak) [inline dts/arm/atmel/sam4s.dtsi]: add uarts when we have driver & yaml for them.

**Review by galak — CHANGES_REQUESTED** [review]
Fixup various license headers to just be SPDX tag

File: dts/arm/atmel/sam4s.dtsi (hunk 4 above)
Comment (fallrisk) [inline dts/arm/atmel/sam4s.dtsi]: OK, I am removing the UART stuff for now.

File: arch/arm/soc/atmel_sam/sam4s/soc.c (hunk 3 above)
Comment (fallrisk) [inline arch/arm/soc/atmel_sam/sam4s/soc.c]: Would the #define go into soc.h?

Comment (fallrisk) [inline arch/arm/soc/atmel_sam/sam4s/soc.c]: Oh! I think I understand what you want. Is it #defines for the bits and this value made up of those?

File: arch/arm/soc/atmel_sam/sam4s/soc.h [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+#define UART_IRQ_FLAGS 0
+
+/* EEFC Register struct */
+#define __EEFC0		((volatile struct __eefc *)EEFC_BANK0_ADDR)
+#define __EEFC1		((volatile struct __eefc *)EEFC_BANK1_ADDR)
+
+/* PMC Register struct */
+#define __PMC		((volatile struct __pmc *)PMC_ADDR)
+
+/* PIO Registers struct */
```
Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/soc.h]: can you trim this back to what is used right now (ie remove PIO, TWI, etc).

File: arch/arm/soc/atmel_sam/sam4s/soc_registers.h [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+/* Enhanced Embedded Flash Controller */
+struct __eefc {
+	u32_t	fmr;	/* 0x00 Flash Mode Register    */
+	u32_t	fcr;	/* 0x04 Flash Command Register */
+	u32_t	fsr;	/* 0x08 Flash Status Register  */
+	u32_t	frr;	/* 0x0C Flash Result Register  */
+};
+
+/* PIO Controller */
```
Comment (galak) [inline arch/arm/soc/atmel_sam/sam4s/soc_registers.h]: remove structs that aren't used right now.

**Review by galak — CHANGES_REQUESTED** [review]
some minor changes, mostly to delete code that isn't currently used.

**Review by galak — APPROVED** [review]

**Pull request merged**